File: python/balanced_strategy.py
"""
Balanced Strategy (Explicit Override)
=====================================
The "balanced" strategy is a fixed-weight override that optimizes the trade-off
between cost, speed, and quality using static weights.

By default (when no strategy is specified), RoutePlex uses auto-routing which
analyzes the prompt to dynamically choose the best model. Use "balanced" only
when you want deterministic fixed-weight routing instead of prompt-based analysis.

Docs: https://routeplex.com/docs/routing-modes
Usage: python balanced_strategy.py
"""

import os
import requests

API_KEY = os.environ.get("ROUTEPLEX_API_KEY")
BASE_URL = "https://api.routeplex.com"

if not API_KEY:
    print("Error: Set ROUTEPLEX_API_KEY environment variable")
    exit(1)


def ask(question: str) -> str:
    """General-purpose question answering with fixed balanced routing."""
    response = requests.post(
        f"{BASE_URL}/api/v1/chat",
        headers={"Authorization": f"Bearer {API_KEY}"},
        json={
            "messages": [{"role": "user", "content": question}],
            "mode": "routeplex-ai",
            "strategy": "balanced",
        },
    )

    if response.status_code != 200:
        print(f"Error {response.status_code}: {response.text}")
        return ""

    data = response.json()
    print(f"  Model: {data['data']['model_used']} | Cost: ${data['data']['usage']['cost_usd']}")
    return data["data"]["output"]


def main():
    questions = [
        "Explain the difference between REST and GraphQL in two sentences.",
        "What are three best practices for database indexing?",
    ]

    for q in questions:
        print(f"\nQ: {q}")
        answer = ask(q)
        print(f"A: {answer}")


if __name__ == "__main__":
    main()

File: typescript/balanced_strategy.ts
/**
 * Balanced Strategy — Fixed-Weight Override (OpenAI SDK)
 *
 * Uses the OpenAI TypeScript SDK with RoutePlex as a drop-in replacement.
 * The "balanced" strategy is a fixed-weight override that optimizes the
 * trade-off between cost, speed, and quality using static weights.
 *
 * By default (when no strategy is specified), RoutePlex uses auto-routing
 * which analyzes the prompt to dynamically choose the best model. Use
 * "balanced" only when you want deterministic fixed-weight routing.
 *
 * Docs: https://routeplex.com/docs/openai-compatibility
 * Usage: npx tsx balanced_strategy.ts
 */

import OpenAI from "openai";

const API_KEY = process.env.ROUTEPLEX_API_KEY;

if (!API_KEY) {
  console.error("Error: Set ROUTEPLEX_API_KEY environment variable");
  process.exit(1);
}

const client = new OpenAI({
  apiKey: API_KEY,
  baseURL: "https://api.routeplex.com/v1",
});

interface Message {
  role: "system" | "user" | "assistant";
  content: string;
}

async function chat(messages: Message[]): Promise<string | null> {
  // Explicit "balanced" override via the X-RoutePlex-Strategy header
  const response = await client.chat.completions.create(
    {
      model: "routeplex-ai",
      messages,
    },
    {
      headers: { "X-RoutePlex-Strategy": "balanced" },
    }
  );

  console.log(`Model: ${response.model}`);
  if (response.usage) {
    console.log(
      `Tokens: ${response.usage.prompt_tokens} in / ${response.usage.completion_tokens} out`
    );
  }

  return response.choices[0].message.content;
}

async function main() {
  const result = await chat([
    { role: "system", content: "You are a concise technical assistant." },
    {
      role: "user",
      content: "Explain the CAP theorem in three sentences.",
    },
  ]);

  console.log(`\nResponse: ${result}`);
}

main();
